contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -0,0 +1,488 @@
+     *  - tilting pool back
+     *  - redeeming OETH
+     */
+    function checkBalance(address _asset)
+        external
+        view
+        virtual
+        override
+        whenNotInBalancerVaultContext
+        returns (uint256 amount)
+    {
+        uint256 bptBalance = _getBalancerPoolTokens();
+
+        /* To calculate the worth of queried asset:
+         *  - assume that all tokens normalized to their ETH value have an equal split balance
+         *    in the pool when it is balanced
+         *  - multiply the BPT amount with the bpt rate to get the ETH denominated amount
+         *    of strategy's holdings
+         *  - divide that by the number of tokens we support in the pool to get ETH denominated
+         *    amount that is applicable to each supported token in the pool.
+         *
+         *    It would be possible to support only 1 asset in the pool (and be exposed to all
+         *    the assets while holding BPT tokens) and deposit/withdraw/checkBalance using only
+         *    that asset. TBD: changes to other functions still required if we ever decide to
+         *    go with such configuration.
+         */
+        amount = (bptBalance.mulTruncate(
+            IRateProvider(platformAddress).getRate()
+        ) / assetsMapped.length);
+
+        /* If the pool asset is equal to (strategy )_asset it means that a rate
+         * provider for that asset exists and that asset is not necessarily
+         * pegged to a unit (ETH).
+         *
+         * Because this function returns the balance of the asset and is not denominated in
+         * ETH units we need to convert the ETH denominated amount to asset amount.
+         */
+        if (toPoolAsset(_asset) == _asset) {
+            amount = amount.divPrecisely(_getRateProviderRate(_asset));
+        }
+    }
+
+    /**
+     * @notice Returns the value of all assets managed by this strategy.
+     * Uses the Balancer pool's rate (virtual price) to convert the strategy's
+     * Balancer Pool Tokens (BPT) to ETH value.
+     * @return value The ETH value
+     *
+     * IMPORTANT if this function is overridden it needs to have a whenNotInBalancerVaultContext
+     * modifier on it or it is susceptible to read-only re-entrancy attack
+     */
+    function checkBalance()
+        external
+        view
+        virtual
+        whenNotInBalancerVaultContext
+        returns (uint256 value)
+    {
+        uint256 bptBalance = _getBalancerPoolTokens();
+
+        // Convert BPT to ETH value
+        value = bptBalance.mulTruncate(
+            IRateProvider(platformAddress).getRate()
+        );
+    }
+
+    /// @notice Balancer Pool Tokens (BPT) in the Balancer pool.
+    function _getBalancerPoolTokens()
+        internal
+        view
+        virtual
+        returns (uint256 balancerPoolTokens)
+    {
+        balancerPoolTokens = IERC20(platformAddress).balanceOf(address(this));
```

Comment:
Noting, I don't think the code in this function is ever used, tested, or included in a deployed contract?

Reply:
This is used by the `checkBalance` functions [here](https://github.com/OriginProtocol/origin-dollar/pull/1697/files#diff-7d59dfab7a512a013f0696d107a97b7b6b0f1e8d656b8baac6642f61e3d84db6R161) and [here](https://github.com/OriginProtocol/origin-dollar/pull/1697/files#diff-7d59dfab7a512a013f0696d107a97b7b6b0f1e8d656b8baac6642f61e3d84db6R208) and both versions (with asset parameter & without) of checkBalance are used in a few places in the [fork test file](https://github.com/OriginProtocol/origin-dollar/pull/1697/files#diff-4ff4ac51088496da8787ba1d1726b09145dc496c6c9a9f9877bf7c47b3fc2522R665)